## Ejercicio 1: Ingesta de datos semiestructurados
+ Utilizando las librerías requests y json , realize una consulta a la API BallDontLie https://www.balldontlie.io/.
+ Mediante esta API, genere una consulta de 100 juegos (busquen el endpoint Get All Games en la página, dentro de Games ) . Puede usar la opción ?per_page=100 en la dirección de la API para lograr esto.
+ Guarde la consulta en una variable, y asegúrese que ésta sea semiestructurada. Consulte los datos y metadatos asociados.
+ Dentro de la lista de datos, comente cuáles son las llaves de cada registro.


In [1]:
import requests
import json
import pandas as pd
import numpy as np

#Bajar datos en un JSON
nba_games = requests.get('https://www.balldontlie.io/api/v1/games?per_page=100').json()


In [2]:
len(nba_games)

2

In [3]:
nba_games['data'][0]

{'id': 47179,
 'date': '2019-01-30T00:00:00.000Z',
 'home_team': {'id': 2,
  'abbreviation': 'BOS',
  'city': 'Boston',
  'conference': 'East',
  'division': 'Atlantic',
  'full_name': 'Boston Celtics',
  'name': 'Celtics'},
 'home_team_score': 126,
 'period': 4,
 'postseason': False,
 'season': 2018,
 'status': 'Final',
 'time': ' ',
 'visitor_team': {'id': 4,
  'abbreviation': 'CHA',
  'city': 'Charlotte',
  'conference': 'East',
  'division': 'Southeast',
  'full_name': 'Charlotte Hornets',
  'name': 'Hornets'},
 'visitor_team_score': 94}

In [4]:
nba_games['data'][0]['home_team']

{'id': 2,
 'abbreviation': 'BOS',
 'city': 'Boston',
 'conference': 'East',
 'division': 'Atlantic',
 'full_name': 'Boston Celtics',
 'name': 'Celtics'}

## Ejercicio 2: Organización de los datos
+ En base a los datos consultados en la API, genere un objeto pd.DataFrame con la siguiente estructura:

season|period|home|home_conference|home_division|visitor|visitor_conference|visitor_division|home_score|visitor_score
---|---|---|---|---|---|---|---|---|---
2018|4|Boston Celtics|East|Atlantic|Charlotte Hornets|East|Southeast|126|94
2018|4|Boston Celtics|East|Atlantic|LA Clippers|West|Pacific|112|123
2018|4|Philadelphia 76ers|East|Atlantic|Denver Nuggets|West|Northwest|117|110
2018|4|Washington Wizards|East|Southeast|Cleveland Cavaliers|East|Central|119|106
2018|4|Sacramento Kings|West|Pacific|Miami Heat|East|Southeast|102|96



+ Para obtener la estructura, sólo puede utilizar operaciones map y comprensiones de lista.


In [ ]:
#defino funcion que procesa cada elemento
def process_nba_games(row): 
    season = row['season']
    period = row['period']
    
    return [season, period]

# mapeo datos usando la funcion
records = list( map(process_nba_games, nba_games['data']))

print(records)

In [10]:
nba_games['data'][0]['home_team']['full_name']

'Boston Celtics'

In [13]:
def myfun(x):
    period = x['period']
    team_name = x['home_team']['full_name']
    return [period, team_name]

records  = list(map(myfun, nba_games['data']))

print(records)

[[4, 'Boston Celtics'], [4, 'Boston Celtics'], [4, 'Philadelphia 76ers'], [4, 'Washington Wizards'], [4, 'Sacramento Kings'], [4, 'Detroit Pistons'], [4, 'New Orleans Pelicans'], [4, 'Indiana Pacers'], [4, 'Atlanta Hawks'], [4, 'Utah Jazz'], [4, 'Chicago Bulls'], [4, 'Dallas Mavericks'], [4, 'Philadelphia 76ers'], [4, 'Sacramento Kings'], [4, 'Milwaukee Bucks'], [4, 'Houston Rockets'], [4, 'Golden State Warriors'], [4, 'Atlanta Hawks'], [4, 'Houston Rockets'], [4, 'Toronto Raptors'], [4, 'Atlanta Hawks'], [4, 'Indiana Pacers'], [4, 'Chicago Bulls'], [4, 'Golden State Warriors'], [4, 'Memphis Grizzlies'], [7, 'Cleveland Cavaliers'], [4, 'Philadelphia 76ers'], [4, 'New Orleans Pelicans'], [4, 'Minnesota Timberwolves'], [4, 'New Orleans Pelicans'], [4, 'Indiana Pacers'], [4, 'Boston Celtics'], [4, 'New York Knicks'], [4, 'Chicago Bulls'], [4, 'Milwaukee Bucks'], [4, 'Cleveland Cavaliers'], [4, 'Orlando Magic'], [4, 'Toronto Raptors'], [4, 'Dallas Mavericks'], [4, 'Portland Trail Blazers']

In [16]:
# creo dataframe
df = pd.DataFrame.from_records(records)
df.columns = ['season', 'home_team']
df.head(20)

season              home_team
0        4         Boston Celtics
1        4         Boston Celtics
2        4     Philadelphia 76ers
3        4     Washington Wizards
4        4       Sacramento Kings
5        4        Detroit Pistons
6        4   New Orleans Pelicans
7        4         Indiana Pacers
8        4          Atlanta Hawks
9        4              Utah Jazz
10       4          Chicago Bulls
11       4       Dallas Mavericks
12       4     Philadelphia 76ers
13       4       Sacramento Kings
14       4        Milwaukee Bucks
15       4        Houston Rockets
16       4  Golden State Warriors
17       4          Atlanta Hawks
18       4        Houston Rockets
19       4        Toronto Raptors

## Ejercicio 3: El efecto de jugar de local
+ Genere una columna en el pd.DataFrame que identifique si el equipo que jugó de local ganó (1) o no (0).
+ Repita el procedimiento para identificar si el equipo que jugó de visitante ganó (1) o no (0).
+ Reporte cuáles son los primeros y últimos 5 equipos en cuanto a desempeño por jugar local.
+ Reporte cuáles son los primeros y últimos 5 equipos en cuanto a desempeño por jugar de visita.


## Ejercicio 4: Obteniendo el porcentaje de ganar local y de visita
+ Genere un nuevo objeto que guarde el porcentaje de juegos ganados como local por equipo.
+ Repita lo mismo para los juegos donde el equipo fue visitante.
+ ¿Qué equipos tienen iguales chances de ganar como visitante o local?
